src/coroutine.rs

```diff
@@ -26,13 +26,44 @@ pub use cancel::CancelHandle;
 
 const COROUTINE_REUSED_ERROR: &str = "cannot reuse already awaited coroutine";
 
+trait CoroutineFuture {
```

Comment:
I am not sure which role the trait plays here? Why not just have a free generic `poll` function?

Reply:
I need a trait to box it.

Reply:
To be more precise, I could not use `Future` anymore because I had to apply `IntoPy` to the `Future` result, see https://github.com/PyO3/pyo3/pull/3610#discussion_r1419234657

Reply:
Indeed. Thinking about the alternatives, this does appear to be a very nice solution. I would suggest having `: Send` as a bound for the trait itself though to make it more obvious why the blanket impl includes that bound as well.

Reply:
done